File: Session6_BFS/same-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean checkSame(TreeNode p, TreeNode q) {
        /* If they are both null then they're equal */
        if (p == null && q == null) return true;
        
        /* If one is defined and one is null, they're not equal */
        if (p == null || q == null) return false;
        
        /* Both nodes are defined and equal */
        if (q.val == p.val) return true;
        
        /* Both defined and not equal */
        return false;
    }
    
    public boolean isSameTree(TreeNode p, TreeNode q) {
        /* Initiate queue */
        Queue<TreeNode> queue = new LinkedList<>();
        /* Enqueue left node, then right node. No need to check root */
        queue.offer(p);
        queue.offer(q);
        
        while (!(queue.isEmpty())) {
            System.out.println(queue);
            
            /* Dequeue left child node, then right child node */
            TreeNode nodeP = queue.poll();
            TreeNode nodeQ = queue.poll();
            
            /* If any nodes are not the same, tree not equal */
            if (!(checkSame(nodeP, nodeQ))) return false;
            
            /**
             * If node is defined, then left and right child
             * won't return an error
             */
            if (nodeP != null) {
                queue.offer(nodeP.right);
                queue.offer(nodeQ.right);
                queue.offer(nodeP.left);
                queue.offer(nodeQ.left);
            }
        }

        /* If this part reached, all nodes checked to be the same. */
        return true;
    }
}

File: Session6_BFS/same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Please use this library if you'd like to implement queues.
# It's more efficient than using arrays! 
from collections import deque

class Solution:
    def checkSame(self, p: TreeNode, q: TreeNode) -> bool:
        """
        Auxilary Function to check if 2 nodes are the same.
        """

        # If they are both null, then they're equal
        if not p and not q:
            return True
        
        # If one is defined and one is null, they're not equal
        if not q or not p:
            return False
        
        # Both nodes are defined and equal
        if q.val == p.val:
            return True
        
        # Both defined and not equal
        return False
    
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        # Inititate queue
        queue = deque()
        # enqueue left node, then right node
        queue.append(p)
        queue.append(q)
        while len(queue) != 0:
            # dequeue left node, then right node
            nodeP = queue.popleft()
            nodeQ = queue.popleft()
            
            # If nodes are not the same, tree not equal
            if not self.checkSame(nodeP, nodeQ):
                return False
            
            # If node is defined, then left and right
            # child won't return an error
            if nodeP:
                queue.append(nodeP.right)
                queue.append(nodeQ.right)
                queue.append(nodeP.left)
                queue.append(nodeQ.left)

        # If this part reached, All nodes checked to be same.  
        return True